### PR TITLE
Allow default grouping option for TOSCA

### LINKED ISSUE #23321: Default detector groups for TOSCA missing
**Original reporter:** Sanghamitra

### Expected behavior

* There should be an option for default detector groups in `ISIS Energy Transfer` in the Indirect Data Reduction interface.

* Detector indices should run from 1

### Actual behavior

* The default grouping was removed in Mantid 3.13 and overlap is not allowed

### Steps to reproduce the behavior

* Open Mantid 3.13
* Open `Indirect` > `Data Reduction`
* Select instrument to be `TOSCA`
* In the `ISIS Energy Transfer` tab, load the file below
[TSC10076.zip](https://github.com/mantidproject/mantid/files/2313393/TSC10076.zip)
* In `Detector Grouping`, under `Mode`, there should be a  `Default` grouping option for when the TOSCA instrument is selected (**Problem 1**)
* Choose `Custom` for `Detector Grouping`
* For range enter `1-70, 71-140, 1-140` fails due to indices problems (**Problem 2**)
* For the range enter `0-70, 71-139, 0-20` fails due to overlap of detectors (**Problem 3**)

### Platforms affected

All

### PR DESCRIPTION
**Pull request by robertapplin**
**Description of work.**
This PR adds an option for default detector groups in ISIS Energy Transfer in the Indirect Data Reduction interface for the TOSCA instrument only. 
This PR also allows spectra numbers to be entered instead of spectra indices in the `Custom` grouping option.
The overlapping of spectra numbers in the `custom` grouping option is also now allowed.

**To test:**
**1.**
- Open Indirect > Data Reduction
- Select instrument to be TOSCA
- In the ISIS Energy Transfer tab, load the file below
[TSC10076.raw](https://files.slack.com/files-pri/T02J4MMEW-FCKLQHRFB/download/tsc10076.raw)
- In Detector Grouping, under Mode, there should be a `Default` grouping option for when the TOSCA instrument is selected (**Problem 1**)
- Click `Run`. The process should run without error and the resultant workspace should contain 3 spectra. 

**2.**
- Use the same file above. In the `Custom` grouping option, enter the following `1-70, 71-140, 1-140` 
- Click 'Run'. The process should run without error. This tests that the spectra numbers (instead of spectra indices) are allowed, and that overlapping groups are allowed. (**Problem 2 and 3**). 
- The resulting workspace should be exactly the same as the output workspace achieved during the default grouping. The algorithm `CompareWorkspaces` could be used to check that this is the case - the table produced by this algorithm should be empty.

Fixes #23321 

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by keeeto — DISMISSED** [review]
This works as described, nearly there.

One thing I notice is that I can set absurd ranges and the reduction tries to run, resulting in Mantid freezing. So for example 
* Set `Detector grouping` to `Custom`
* Choose a range of `1-60000000`
* Hit Run

Can you add a check before reduction to ensure that the range entered makes sense with respect to the data provided.

**Review by keeeto — APPROVED** [review]

File: qt/scientific_interfaces/Indirect/ISISEnergyTransfer.cpp [hunk 1]
```diff
@@ -48,6 +50,17 @@ std::string createDetectorGroupingString(std::size_t numberOfDetectors,
...
 }
 
+std::vector<std::size_t>
+getCustomGroupingNumbers(std::string const &customString) {
+  std::vector<std::string> customGroupingStrings;
+  std::vector<std::size_t> customGroupingNumbers;
+  // Get the numbers from customString and store them in customGroupingStrings
+  boost::split(customGroupingStrings, customString, boost::is_any_of(" ,-+:"));
+  for (const auto &string : customGroupingStrings)
+    if (string != "")
```
Comment (gvardany) [inline qt/scientific_interfaces/Indirect/ISISEnergyTransfer.cpp]: `!string.empty()` is preferable for this check.

File: qt/scientific_interfaces/Indirect/ISISEnergyTransfer.cpp [hunk 2]
```diff
@@ -230,14 +243,36 @@ bool ISISEnergyTransfer::validate() {
...
   if (m_uiForm.cbGroupingOptions->currentText() == "File") {
     if (!m_uiForm.dsMapFile->isValid())
       return "Mapping file is invalid.";
-  }
-  if (m_uiForm.cbGroupingOptions->currentText() == "Custom") {
-    if (m_uiForm.leCustomGroups->text() == "")
+  } else if (m_uiForm.cbGroupingOptions->currentText() == "Custom") {
+    const std::string customString =
+        m_uiForm.leCustomGroups->text().toStdString();
+    if (customString == "")
```
Comment (gvardany) [inline qt/scientific_interfaces/Indirect/ISISEnergyTransfer.cpp]: Same applies here.

File: qt/scientific_interfaces/Indirect/ISISEnergyTransfer.cpp (hunk 1 above)
Comment (robertapplin) [reply qt/scientific_interfaces/Indirect/ISISEnergyTransfer.cpp]: That is better, thanks!

**Pull request merged**